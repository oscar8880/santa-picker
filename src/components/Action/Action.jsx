import styles from './Action.module.scss';
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';

const Action = ({
  tagName = 'button',
  appearance = 'button-primary',
  fullWidth = false,
  children,
}) => {
  const Tag = tagName === 'Link' ? Link : tagName;
  return (
    <Tag
      className={classNames(
        styles.Action,
        styles[`Action__Appearance--${appearance}`],
        {
          [styles.Action__FullWidth]: fullWidth,
        },
      )}
    >
      {children}
    </Tag>
  );
};

Action.propTypes = {
  tagName: PropTypes.oneOf(['button', 'a', 'link']),
  appearance: PropTypes.oneOf([
    'button-primary',
    'button-secondary',
    'button-primary-outline',
    'button-secondary-outline',
    'a',
    'link',
  ]),
  fullWidth: PropTypes.bool,
  children: PropTypes.string,
};

export default Action;
