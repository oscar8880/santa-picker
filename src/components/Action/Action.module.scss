@import "../../scss/functions";
@import "../../scss/variables";

.Action {
  align-items: center;
  -webkit-appearance: none;
  border: 0;
  cursor: pointer;
  position: relative;
  text-decoration: none;
  transition: $transition-default;
}

.Action__Appearance--button-primary,
.Action__Appearance--button-secondary,
.Action__Appearance--button-primary-outline,
.Action__Appearance--button-secondary-outline {
  border-style: solid;
  border-width: 1px;
  display: inline-flex;
  font-family: $font-family-base;
  font-weight: bold;
  justify-content: center;
  line-height: rem(22px);
  padding: 10px 40px;
  text-align: center;
}

.Action__Appearance--button-primary {
  background-color: $primary-colour;
  border-color: $primary-colour;
  color: $white;

  &:hover:not(:disabled),
  &:focus:not(:disabled),
  &:active:not(:disabled) {
    background-color: darken($primary-colour, 5%);
    border-color: darken($primary-colour, 5%);
  }

  &:disabled {
    cursor: default;
    opacity: 0.5;
  }
}

.Action__Appearance--button-secondary {
  background-color: $secondary-colour;
  border-color: $secondary-colour;
  color: $white;

  &:disabled {
    cursor: default;
    opacity: 0.5;
  }

  &:hover:not(:disabled),
  &:focus:not(:disabled),
  &:active:not(:disabled) {
    background-color: darken($secondary-colour, 5%);
    border-color: darken($secondary-colour, 5%);
  }
}

.Action__Appearance--button-primary-outline,
.Action__Appearance--button-secondary-outline {
  background-color: transparent;
}

.Action__Appearance--button-primary-outline {
  border-color: $primary-colour;
  color: $primary-colour;

  &:hover,
  &:focus,
  &:active {
    background-color: $primary-colour;
    color: $white;
  }
}

.Action__Appearance--button-secondary-outline {
  border-color: $secondary-colour;
  color: $secondary-colour;

  &:hover,
  &:focus,
  &:active {
    background-color: $secondary-colour;
    color: $white;
  }
}

.Action__Appearance--a {
  background-color: transparent;
  color: #007aff;

  &:hover,
  &:focus,
  &:active {
    color: darken(#007aff, 5%);
  }
}

.Action__Appearance--none {
  appearance: none;
  background-color: transparent;
  border: 0;
  color: inherit;
  cursor: pointer;
  display: inline-block;
  font-size: inherit;
  line-height: inherit;
  padding: 0;
  text-align: inherit;
}

.Action__FullWidth {
  width: 100%;
}
