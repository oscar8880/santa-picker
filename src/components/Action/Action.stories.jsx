import React from 'react';
import { MemoryRouter } from 'react-router';
import Action from './Action';

export default {
  component: Action,
  title: 'Action',
  argTypes: {
    appearance: {
      control: {
        type: 'inline-radio',
        options: [
          'button-primary',
          'button-secondary',
          'button-primary-outline',
          'button-secondary-outline',
          'a',
          'none',
        ],
      },
    },
    tagName: {
      control: {
        type: 'inline-radio',
        options: ['button', 'a', 'Link'],
      },
    },
    children: { control: 'none' },
  },
};

export const Default = (args) => {
  return (
    <>
      <MemoryRouter>
        <Action {...args}>Click me</Action>
      </MemoryRouter>
    </>
  );
};
