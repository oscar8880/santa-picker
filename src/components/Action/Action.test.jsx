import React from 'react';
import { render } from '@testing-library/react';

import Action from './Action';

test('matches snapshot', async () => {
  const { asFragment } = render(<Action />);

  expect(asFragment()).toMatchSnapshot();
});

test('matches snapshot with props set', async () => {
  const { asFragment } = render(
    <Action appearance="button-secondary-outline" tagName="a" />,
  );

  expect(asFragment()).toMatchSnapshot();
});
