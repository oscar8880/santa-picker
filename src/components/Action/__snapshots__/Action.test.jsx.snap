// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="Action Action__Appearance--button-primary"
  />
</DocumentFragment>
`;

exports[`matches snapshot with props set 1`] = `
<DocumentFragment>
  <a
    class="Action Action__Appearance--button-secondary-outline"
  />
</DocumentFragment>
`;
